File: backend/pkg/database/models/organization.go
package models

import (
	"encoding/json"
	"time"

	"backend/pkg/core/organization"

	"github.com/google/uuid"
	"gorm.io/datatypes"
)

// OrganizationModel is the GORM model for organizations
type OrganizationModel struct {
	ID              uuid.UUID      `json:"id" gorm:"type:uuid;primaryKey;default:uuid_generate_v4()"`
	Name            string         `json:"name" gorm:"not null"`
	Slug            string         `json:"slug" gorm:"uniqueIndex;not null"`
	Bio             *string        `json:"bio"`
	LogoURL         *string        `json:"logo_url"`
	WebsiteURL      *string        `json:"website_url"`
	EmailPublic     *string        `json:"email_public"`
	SocialLinks     datatypes.JSON `json:"social_links" gorm:"type:jsonb;default:'{}'"`
	MetaDescription *string        `json:"meta_description"`
	CreatedAt       time.Time      `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt       time.Time      `json:"updated_at" gorm:"autoUpdateTime"`
}

func (OrganizationModel) TableName() string {
	return "organization"
}

// ToCore converts the GORM model to the domain type
func (m *OrganizationModel) ToCore() *organization.Organization {
	var socialLinks map[string]interface{}
	if m.SocialLinks != nil {
		_ = json.Unmarshal(m.SocialLinks, &socialLinks)
	}

	return &organization.Organization{
		ID:              m.ID,
		Name:            m.Name,
		Slug:            m.Slug,
		Bio:             m.Bio,
		LogoURL:         m.LogoURL,
		WebsiteURL:      m.WebsiteURL,
		EmailPublic:     m.EmailPublic,
		SocialLinks:     socialLinks,
		MetaDescription: m.MetaDescription,
		CreatedAt:       m.CreatedAt,
		UpdatedAt:       m.UpdatedAt,
	}
}

// OrganizationModelFromCore creates a GORM model from the domain type
func OrganizationModelFromCore(o *organization.Organization) *OrganizationModel {
	var socialLinks datatypes.JSON
	if o.SocialLinks != nil {
		socialLinks, _ = datatypes.NewJSONType(o.SocialLinks).MarshalJSON()
	}

	return &OrganizationModel{
		ID:              o.ID,
		Name:            o.Name,
		Slug:            o.Slug,
		Bio:             o.Bio,
		LogoURL:         o.LogoURL,
		WebsiteURL:      o.WebsiteURL,
		EmailPublic:     o.EmailPublic,
		SocialLinks:     socialLinks,
		MetaDescription: o.MetaDescription,
		CreatedAt:       o.CreatedAt,
		UpdatedAt:       o.UpdatedAt,
	}
}

File: backend/src/database/models/organization.rs
use chrono::{DateTime, Utc};
use diesel::{Identifiable, Insertable, Queryable, Selectable};
use serde_json::Value;
use uuid::Uuid;

use crate::{core::organization::Organization, error::AppError, schema::organization};

#[derive(Debug, Clone, Queryable, Selectable, Identifiable)]
#[diesel(table_name = organization)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct OrganizationRow {
    pub id: Uuid,
    pub name: String,
    pub slug: String,
    pub bio: Option<String>,
    pub logo_url: Option<String>,
    pub website_url: Option<String>,
    pub email_public: Option<String>,
    pub social_links: Option<Value>,
    pub meta_description: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Insertable)]
#[diesel(table_name = organization)]
pub struct NewOrganizationRow<'a> {
    pub id: Uuid,
    pub name: &'a str,
    pub slug: &'a str,
    pub bio: Option<&'a str>,
    pub logo_url: Option<&'a str>,
    pub website_url: Option<&'a str>,
    pub email_public: Option<&'a str>,
    pub social_links: Option<Value>,
    pub meta_description: Option<&'a str>,
}

impl TryFrom<OrganizationRow> for Organization {
    type Error = AppError;

    fn try_from(row: OrganizationRow) -> Result<Self, Self::Error> {
        let social_links = match row.social_links {
            Some(Value::Object(values)) => Some(values.into_iter().collect()),
            Some(_) => None,
            None => None,
        };
        Ok(Self {
            id: row.id,
            name: row.name,
            slug: row.slug,
            bio: row.bio,
            logo_url: row.logo_url,
            website_url: row.website_url,
            email_public: row.email_public,
            social_links,
            meta_description: row.meta_description,
            created_at: row.created_at,
            updated_at: row.updated_at,
        })
    }
}
